rust/kernel/src/sync.rs

```diff
@@ -0,0 +1,139 @@
+// SPDX-License-Identifier: GPL-2.0
+
+use core::cell::UnsafeCell;
+use core::ops::{Deref, DerefMut};
+use core::fmt;
+
+use crate::bindings;
+
+/*#[cfg(CONFIG_DEBUG_SPINLOCK)]
+compile_error!("CONFIG_DEBUG_SPINLOCK not yet implemented");
+
+#[cfg(CONFIG_DEBUG_MUTEXES)]
+compile_error!("CONFIG_DEBUG_MUTEX not yet implemented");*/
+
+pub struct Mutex<T: ?Sized> {
+    lock: bindings::mutex,
+    data: UnsafeCell<T>,
+}
+
+impl<T> Mutex<T> {
+    /// Create a new Mutex
+    pub fn new(data: T, _name: &'static str) -> Self {
+
+        let lock = bindings::mutex::default();
+        // TODO: write mutex debug traces like .magic.
+        // TODO: use name in the debug version
+
+        Self {
+            data: UnsafeCell::new(data),
+            lock
+        }
+    }
+}
+
+impl<T: ?Sized> Mutex<T> {
+    /// acquire a lock on the mutex
+    #[cfg(CONFIG_DEBUG_LOCK_ALLOC)]
+    pub fn lock<'a>(&'a self) -> MutexGuard<'a, T> {
+        unsafe {
+            //bindings::mutex_lock(&self.lock as *const bindings::mutex as *mut bindings::mutex);
+            bindings::mutex_lock_nested(&self.lock as *const bindings::mutex as *mut bindings::mutex, 0);
+        }
+        MutexGuard { inner: &self }
+    }
+
+    /// acquire a lock on the mutex
+    #[cfg(not(CONFIG_DEBUG_LOCK_ALLOC))]
+    pub fn lock<'a>(&'a self) -> MutexGuard<'a, T> {
+        unsafe {
+            bindings::mutex_lock(&self.lock as *const bindings::mutex as *mut bindings::mutex);
+        }
+        MutexGuard { inner: &self }
+    }
+
+    /// try to acquire the lock, returns none on failure
+    pub fn try_lock<'a>(&'a self) -> Option<MutexGuard<'a, T>> {
+        let ret = unsafe { bindings::mutex_trylock(&self.lock as *const bindings::mutex as *mut bindings::mutex) };
```

Comment:
I suggest going with the same names as in the C side

Reply:
I named it try_lock as the std Mutex has it like this

Reply:
What why do we want to go? More rust standards, or c side standards?

Reply:
There are advantages to both ways, so it is not clear. I guess we will see as we go what is best.

I am inclined to follow the kernel names. There are two main reasons for this. One, for consistency, for easy `grep`ing and to make it as familiar as possible for current kernel developers, etc. But the other reason is that in many cases the kernel API does not really look or behave as the typical userspace counterparts. For instance, memory allocation requires flags vs. `malloc()` or printing (the kernel log with levels etc. vs. `println!`/`stdout/err`). They also may be unsafe (we should provide safe APIs too wherever applicable, but I am pretty sure people will demand having access to unsafe ones for performance in some cases).